app/javascript/packages/document-capture/components/in-person-location-post-office-search-step.tsx

```diff
@@ -77,15 +86,82 @@ const requestUspsLocations = async (address: LocationQuery): Promise<FormattedLo
   return formatLocation(response);
 };
 
-function InPersonLocationPostOfficeSearchStep({ onChange, toPreviousStep, registerField }) {
+const ADDRESS_SEARCH_URL = '/api/addresses';
+
+function requestAddressCandidates(unvalidatedAddressInput: string): Promise<Location[]> {
+  return request<Location[]>(ADDRESS_SEARCH_URL, {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    json: { address: unvalidatedAddressInput },
+  });
+}
+
+function useUspsLocations() {
+  // raw text input that is set when user clicks search
+  const [addressQuery, setAddressQuery] = useState('');
+  const [fieldValidation, setFieldValidation] = useState<RefObject<HTMLFormElement>>();
   const { t } = useI18n();
+  const handleAddressSearch = useCallback((event, unvalidatedAddressInput, validatedFieldRef) => {
+    event.preventDefault();
+    validatedFieldRef.current?.setCustomValidity('');
+    validatedFieldRef.current?.reportValidity();
+    setFieldValidation(validatedFieldRef);
+    if (unvalidatedAddressInput === '') {
+      return;
+    }
+
+    setAddressQuery(unvalidatedAddressInput);
+  }, []);
+
+  // sends the raw text query to arcgis
+  const { data: addressCandidates, isLoading: isLoadingCandidates } = useSWR(
+    [addressQuery],
+    () => (addressQuery ? requestAddressCandidates(addressQuery) : null),
+    { keepPreviousData: true },
+  );
+
+  // sets the arcgis-validated address object
   const [foundAddress, setFoundAddress] = useState<LocationQuery | null>(null);
+  const handleFoundAddress = useCallback((address) => {
+    setFoundAddress({
+      streetAddress: address.street_address,
+      city: address.city,
+      state: address.state,
+      zipCode: address.zip_code,
+      address: address.address,
+    });
+  }, []);
+
+  useEffect(() => {
+    if (addressCandidates) {
+      const bestMatchedAddress = addressCandidates[0];
+      const validity = bestMatchedAddress ? '' : t('in_person_proofing.body.location.inline_error');
+      fieldValidation?.current?.setCustomValidity(validity);
+      fieldValidation?.current?.reportValidity();
+      bestMatchedAddress && handleFoundAddress(bestMatchedAddress);
+    }
+  }, [addressCandidates]);
```

Comment:
These can be combined and made safer.

```suggestion
  useEffect(() => {
    if (addressCandidates?.[0]) {
      const bestMatchedAddress = addressCandidates[0];
      setFoundAddress({
        streetAddress: bestMatchedAddress.street_address,
        city: bestMatchedAddress.city,
        state: bestMatchedAddress.state,
        zipCode: bestMatchedAddress.zip_code,
        address: bestMatchedAddress.address,
      });
    } else if (addressCandidates) {
      fieldValidation.current?.setCustomValidity(
        t('in_person_proofing.body.location.inline_error')
      )
      fieldValidation.current?.reportValidity();
    }
  }, [addressCandidates]);
```